### PR TITLE
`Architecture`: Migrate to micro services: create JHipster gateway

### PR DESCRIPTION
**Pull request by MerlinMehmed**
### Checklist
Not applicable

### Motivation and Context

This PR is a step further towards the migration of Artemis to microservices. It adds an API Gateway between the client and the server. The role of the API Gateway is to act as an entrance to the microservices. It also does HTTP routing and load balancing for all microservices.

### Description

The PR adds an API Gateway, generated by JHIpster and modified for Artemis' requirements. The client is packaged in the gateway while the server code is packaged in a separate war file. The server can be seen as a big microservice. 
Both the gateway and the Artemis server application register themselves in a registry which helps the gateway to know which microservices are up. This way the gateway knows where to redirect the requests to.

**The Gateway's configurations(dependency versions, security configurations, etc.) are expected to be the same as the ones for Artemis. There is one difference that is related to the security configurations. They will look different than Artemis in the aspect of WebFlux.**
JHipster has added support for the Spring WebFlux framework which complements Spring MVC and allows to use reactive programming in Java applications. WebFlux is a non-blocking framework built to take an advantage of multi-core processors and handle massive numbers of concurrent connections with small number of threads. This helps the system to scale with fewer hardware resources. To compare, Spring MVC uses synchronous blocking architecture with one-reques-per-thread model. 
The Gateway was generated by JHipster using WebFlux by default. The Gateway is the entrypoint to our server and every single request goes through it. There is a general rule of thumb that WebFlux will make difference compared to Spring MVC if you have >500 requests/second which will most likely be fulfilled by Artemis when running exams, quizzes for large courses (i.e. EIST). There is no change to any of the configurations or REST endpoints in Artemis. WebFlux is only enabled for the Gateway project. The Gateway redirects all requests from the client to the Artemis application which still uses Spring MVC. The difference is that when the request is redirected the gateway can handle another request with the same thread instead of waiting for an answer. Once Artemis sends back response the Gateway will be notified about that and will finalize the request and send it to the client.

### Steps for Testing

This PR should be tested locally and there are two ways to do it - with or without Docker. The one without Docker requires you to clone the JHipster Registry while the one with Docker does that for you. The second one with provided for people who don't want to setup Docker.

**Using Docker**
1. Update application.yml file and set values: 
```
jhipster.registry.password=admin
eureka.client.enabled=true
```
2. Update application-dev.yml file and set values: 
`eureka.client.enabled=true`
3. Run the server application as you usually do, using IntelliJ or gradlew
4. Navigate to /gateway folder in the file system
5. Build the gateway image using
```
./gradlew :gateway:bootJar -Pprod jibDockerBuild
```
6. Run the Docker compose file to start the gateway and the registry
```
docker-compose -f src/main/docker/app.yml up -d
```

**Without Docker**
1. Update application.yml file and set values: 
```
jhipster.registry.password=admin
eureka.client.enabled=true
```
2. Update application-dev.yml file and set values: 
`eureka.client.enabled=true`
3. Run the server application as you usually do, using IntelliJ or gradlew
4. Clone the JHipster Registry from the following repository: [https://github.com/jhipster/jhipster-registry](https://github.com/jhipster/jhipster-registry)
5. Run the Registry:
```
./mvnw -Pdev,webapp
```
6. Navigate to /gateway folder in the file system
7. Start the gateway application using
```
./gradlew
```
8. Continue with the common steps

**Common steps**
1. Navigate to [http://localhost:8761/](http://localhost:8761/)
2. Login with credentials admin / admin
3. You should see the Artemis application and the Gateway registered
Screenshot from 2021-09-04 18-35-14
4. Navigate to [http://localhost:8089/](http://localhost:8089/)
5. You should see the Artemis login page
6. Try to login and it should work
7. Try to do other actions: participate in exercise, create course or an exercise, create user

### Review Progress

Most of the code was auto-generated by JHipster, therefore I suppose that the review would be fast. Although, since there are many new files, I splitted the code review to several parts
- Code Review
  - [x] Review gateway/build.gradle, gateway/gradle.properites, gateway/settings.gradle to make sure that the dependency versions of common packages match with the ones in Artemis (1)
  - [x]  Review gateway/build.gradle, gateway/gradle.properites, gateway/settings.gradle to make sure that the dependency versions of common packages match with the ones in Artemis (2)
  - [x] Review security configurations(SecurityConfiguration class and de.tum.in.www1.artemis.security package in the _gateway_ folder), make sure the configurations will work similarly as the ones in Artemis (1)
  - [x] Review security configurations(SecurityConfiguration class and de.tum.in.www1.artemis.security package in the _gateway_ folder), make sure the configurations will work similarly as the ones in Artemis (2)
  - [x] Check the routes routing under spring.cloud.routes configuration in application.yml file in the _gateway_ folder. Make sure that there are no missing routes that should be forwarded to the Artemis server (1)
  - [x] Check the routes routing under spring.cloud.routes configuration in application.yml file in the _gateway_ folder. Make sure that there are no missing routes that should be forwarded to the Artemis server (2)
  - [x] Review build server in war (gradle/war.gradle and build.gradle files) (1)
  - [x] Review build server in war (gradle/war.gradle and build.gradle files) (2)

- Manual Tests
  - [x] Test 1
  - [x] Test 2

### CODE REVIEW

**Review by lschlesinger — DISMISSED** [review]
I followed the testing steps for docker and everything worked as expected, I could login, create a text exercise and a lecture. I can approve the functionality!
Maybe you could add this little flag local/dev, I did not see it.
Bildschirmfoto 2021-09-13 um 19 33 17

**Review by Malyuk-A — DISMISSED** [review]
Tested locally during a testing session. Works as described 👍

**Review by imstreetspirit — DISMISSED** [review]
tested locally on Windows machine during testing session. I followed the steps with docker setup. I logged in Artemis successfully and created an exercise, worked as described.

**Review by ndricimrr — COMMENTED** [review]
Tested during testing session. Could create an exercise properly no issues. JHipster route also working fine. 
One small note I get some small console log when opening  **course-management/1/exercises**.
`ERROR TypeError: Cannot read properties of undefined (reading 'includes') at feature-toggle.service.ts:79 ` 
Not sure if we use it in our PR, but it would be nice if you could double check that functonality works. 
Other then that all good by me 👍

File: gateway/gradle.properties [hunk 1]
```diff
@@ -0,0 +1,34 @@
+rootProject.name=gateway
+profile=dev
+
+# Build properties
+node_version=14.16.1
+yarn_version=1.22.10
+
```
Comment (Malyuk-A) [inline gateway/gradle.properties:7]: What is the reason why hibernate is not present in the gradle.properties for the gateway (but present for artemis?)

File: gateway/settings.gradle [hunk 2]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (Malyuk-A) [inline gateway/settings.gradle:1]: Similar to above why are the ids for liquibase, eclipse, etc. "missing" Or is it fine?
image [settings.gradle]

**Review by Malyuk-A — COMMENTED** [review]
Started looking at the code (after discussing what should be reviewed in the first place)
I will continue now, but I wanted to provide some comments already.

File: gateway/src/main/java/de/tum/in/www1/artemis/security/jwt/JWTFilter.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package de.tum.in.www1.artemis.security.jwt;
```
Comment (Malyuk-A) [inline gateway/src/main/java/de/tum/in/www1/artemis/security/jwt/JWTFilter.java:1]: Similar to above, maybe it is not as relevant but it is still in the security directory.
Here is the diff https://www.diffchecker.com/SpzPCbmj
It differs strongly.

File: gateway/src/main/java/de/tum/in/www1/artemis/security/jwt/TokenProvider.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
+package de.tum.in.www1.artemis.security.jwt;
```
Comment (Malyuk-A) [inline gateway/src/main/java/de/tum/in/www1/artemis/security/jwt/TokenProvider.java:1]: Similar as above:
Here is the diff: https://www.diffchecker.com/kOMSbyho
[TokenProvider.java]

File: gateway/src/main/resources/config/application-dev.yml [hunk 5]
```diff
@@ -0,0 +1,119 @@
+# ===================================================================
```
Comment (Malyuk-A) [inline gateway/src/main/resources/config/application-dev.yml:1]: (simmilar)
diff link: https://www.diffchecker.com/RfyVjNKn

File: gateway/src/main/resources/config/application-prod.yml [hunk 6]
```diff
@@ -0,0 +1,140 @@
+# ===================================================================
```
Comment (Malyuk-A) [inline gateway/src/main/resources/config/application-prod.yml:1]: Similar
Diff Link: https://www.diffchecker.com/Yoyw1sPB

File: gateway/build.gradle [hunk 7]
```diff
@@ -0,0 +1,201 @@
+buildscript {
```
Comment (Malyuk-A) [inline gateway/build.gradle:1]: Diff Link: https://www.diffchecker.com/ZFnTumDd

**Review by Malyuk-A — COMMENTED** [review]
I looked through the mentioned points/files in the code review section of the description.
Due to my scarce experience with gradle, security, and setup files, I mostly looked at differences but if there were large numbers of "missing" code blocks I can not really tell if this is problematic or not. During the testing session I installed it localy and it worked successfully though.

To help further discussions and reviers I created comments with links in them that show the differences between the artemis and gateway files. I hope they will be useful.

File: gateway/settings.gradle (hunk 2 above)
Comment (MerlinMehmed) [reply gateway/settings.gradle:1]: Same for liquibase as for hibernate.
I am not sure what the eclipse plugin is used for. I can add it, it's not a problem. Are you familiar what it is used for?

File: gateway/gradle.properties (hunk 1 above)
Comment (MerlinMehmed) [reply gateway/gradle.properties:7]: We do not have a db connection in the gateway, neither db entities and repositories we work with, therefore it is not used and not needed.

File: gateway/src/main/resources/config/application-dev.yml (hunk 5 above)
Comment (MerlinMehmed) [reply gateway/src/main/resources/config/application-dev.yml:1]: I think here the main problem is with the space counts, good catch

File: gateway/settings.gradle (hunk 2 above)
Comment (Malyuk-A) [reply gateway/settings.gradle:1]: Unfortunately not really, I assume the eclipse plugin is used to maybe support Orion or similar things, so you can use Artemis directly in your IDE but I am not sure.

**Review by julian-christl — DISMISSED** [review]
Great job! Now it's even only 78 files.
So from my side code review is finished. The codebase looks good to me. 🎉

**Review by Olimbia23 — DISMISSED** [review]
I had a look into some parts of your code. 
Nice work @MerlinMehmed 
The changes look good.
Approve 👍

File: gateway/src/test/java/de/tum/in/www1/artemis/gateway/web/rest/TestUtil.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        return new NumberMatcher(number);
+    }
+
+    /**
+     * Verifies the equals/hashcode contract on the domain object.
+     *
+     * @param clazz
+     * @param <T>
+     * @throws Exception
+     */
```
[the remark this answers is not in the document]
Comment (MerlinMehmed) [reply gateway/src/test/java/de/tum/in/www1/artemis/gateway/web/rest/TestUtil.java:165]: you are right, but I found out the method is not used, therefore I deleted it

File: gateway/src/test/java/de/tum/in/www1/artemis/gateway/ArchTest.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import com.tngtech.archunit.core.importer.ImportOption;
+
+class ArchTest {
+
+    @Test
+    void servicesAndRepositoriesShouldNotDependOnWebLayer() {
+        JavaClasses importedClasses = new ClassFileImporter().withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS).importPackages("de.tum.in.www1.artemis.gateway");
+
+        noClasses().that().resideInAnyPackage("de.tum.in.www1.artemis.gateway.service..").or().resideInAnyPackage("de.tum.in.www1.artemis.gateway.repository..").should()
+                .dependOnClassesThat().resideInAnyPackage("..de.tum.in.www1.artemis.gateway.web..").because("Services and repositories should not depend on web layer")
```
[the remark this answers is not in the document]
Comment (MerlinMehmed) [reply gateway/src/test/java/de/tum/in/www1/artemis/gateway/ArchTest.java:18]: done

File: gateway/src/test/java/de/tum/in/www1/artemis/gateway/security/jwt/TokenProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    @Test
+    void testKeyIsSetFromBase64SecretWhenSecretIsEmpty() {
+        final String base64Secret = "fd54a45s65fds737b9aafcb3412e07ed99b267f33413274720ddbb7f6c5e64e9f14075f2d7ed041592f0b7657baf8";
+        JHipsterProperties jHipsterProperties = new JHipsterProperties();
+        jHipsterProperties.getSecurity().getAuthentication().getJwt().setBase64Secret(base64Secret);
+
+        TokenProvider tokenProvider = new TokenProvider(jHipsterProperties);
+
+        Key key = (Key) ReflectionTestUtils.getField(tokenProvider, "key");
+        assertThat(key).isNotNull().isEqualTo(Keys.hmacShaKeyFor(Decoders.BASE64.decode(base64Secret)));
```
[the remark this answers is not in the document]
Comment (MerlinMehmed) [reply gateway/src/test/java/de/tum/in/www1/artemis/gateway/security/jwt/TokenProviderTest.java:112]: you are right :thumbsup:

File: gateway/src/main/java/de/tum/in/www1/artemis/gateway/web/rest/GatewayResource.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            String path = predicate.substring(predicate.indexOf("[") + 1, predicate.indexOf("]"));
+            routeVM.setPath(path);
+            String serviceId = route.getId().substring(route.getId().indexOf("_") + 1).toLowerCase();
+            routeVM.setServiceId(serviceId);
+            // Exclude gateway app from routes
+            if (!serviceId.equalsIgnoreCase(appName)) {
+                routeVM.setServiceInstances(discoveryClient.getInstances(serviceId));
+                routeVMs.add(routeVM);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (MerlinMehmed) [reply gateway/src/main/java/de/tum/in/www1/artemis/gateway/web/rest/GatewayResource.java:59]: I think that in this case we shouldn't use toStream. The gateway generated by JHipster is Spring Cloud Gateway which should be non-blocking and toStream is a blocking operation. 
If I try to use toStream I get an Internal server error with this message: `Iterating over a toIterable() / toStream() is blocking, which is not supported in thread parallel`.

**Review by julian-christl — DISMISSED** [review]
Changes look good

**Review by julian-christl — APPROVED** [review]
There are now three Codacy warnings. Could you please check them out?

**Review by imstreetspirit — APPROVED** [review]
I tested again on Windows machine by following steps with Docker, everything went well, could not spot any issue on Artemis.

**Review by imstreetspirit — APPROVED** [review]
re-approve, still works as expected

**Review by julian-christl — DISMISSED** [review]
Reapprove

**Review by Olimbia23 — DISMISSED** [review]
Thank you for addressing my suggestions @MerlinMehmed
Code changes look good.
Approve 👍

**Review by argertboja — APPROVED** [review]
I tested the PR on my windows machine using Docker. I played around with different cases, creating exercises, participating in them, assessing them and I didn't encounter any errors!

Well done @MerlinMehmed , you did a great job 👍🏼

**Review by evgeny1984 — DISMISSED** [review]
Nice work!

**Review by jpbernius — CHANGES_REQUESTED** [review]
Changes look good. One minor thing, probably due to code formatting:

**Review by imstreetspirit — DISMISSED** [review]
re-approve, only small strings changed after my last review and it looks good.

**Review by julian-christl — DISMISSED** [review]
Code Reapproval

Changes don't affect functionality => No retesting required

**Pull request merged**